### ISSUE TITLE
Proposed new documentation workflow

### ISSUE THREAD

**Issue by mvertens** [ESMCI/cime#1088]
I propose to have the CIME documentation for the user's guide, data models and driver become a part of CIME in a doc directory at the top level.  This doc/ directory would have all of the sphinx hooks in it and would be easily able to generate the html pages if the contributor had sphinx installed on their system. I have prototyped this in the branch mvertens/doc in esmci. 
ASSUMING you have sphinx, all you would do is 
- edit the relevant doc/ files
- issue make html for doc/
- immediately view the result in your browser 

I think that this approach has several advantages over our current approach. 
- It is much easier to modify the documentation (in my case using emacs in rst mode). 
- Changes to CIME could then easily have the documentation update as part of a PR. This would make it much easier to have the documentation keep up to date with changes in CIME. If it is in two places, experience has been that it is really never updated frequently. In fact we could put in an additional field in the PR template asking if the documentation was updated as part of the PR.
- I found working with the current wiki system to be clumsy and time-consuming. I was able to put in changes and view them quickly and the resulting new documentation was immediately translated into the form we would have for publication.

**Comment by rljacob** [comment]
You can always clone cime.wiki and edit rst files with emacs or your favorite editor and then generate html locally.   You don't *have* to edit through the web interface.  But the web interface is great for making quick changes.

**Comment by mvertens** [comment]
Yes you can do this. The format in the wiki was very klunky and not easily
integrated into sphinx. What I have done in my branch is integrate it
nicely into a sphinx workflow that will let us work very quickly with a
publishable format and potentially keep this up to date with cime changes.
Could you please look at what I have - or I could show it tomorrow.
I am still trying to figure out how to easily publish this to the github
pages.

**Comment by gold2718** [comment]
As the person who initially set up the gh-pages branch to publish, I thought going straight from the wiki would be an easy path. However, I could never get the publication to go cleanly in that direction (I recently tried again when @rljacob asked me to in issue #1026). The publication workflow from the main repo is much cleaner and being free from the wiki restrictions means more professional documentation.
I still think we should use the wiki for some internal CIME development work (TBD) while material which has moved to the doc directory has links from the wiki (so the wiki is still a good way to quickly access CIME documentation).
I also like the idea of documentation updates being part of CIME PRs.

**Comment by billsacks** [comment]
With the caveat that I haven't done much documentation yet myself, @mvertens 's suggestion sounds good to me.

Would this https://gist.github.com/cobyism/4730490 be useful to address this??? (I haven't read it in detail, but it looks like it's trying to solve the same problem):

> I am still trying to figure out how to easily publish this to the github pages.

**Comment by mvertens** [comment]
I have migrated the wiki pages for the users' guide, data model guide and driver guide to the sphinx workflow I proposed. You can see this in http://esmci.github.io/cime/doc/build/html/index.html.
All of this work is in the gh-pages branch in esmci. For now, and until we have a solid version of the documentation, I propose that we work off of this branch. We can continue in the meantime to discuss how we want to manage the documentation in the longer term.
We should also check out the documentation that Bill is pointing to.

**Comment by gold2718** [comment]
@billsacks, We could do something like that suggested in the gist (summary: just push the documentation sub-directory to the gh-pages branch).
For now, the root index.html just redirects to the correct HTML page. Let's keep your suggestion in mind when we decide what to do with the reasonably mature documentation.
If we use this method, documentation would always be edited in master (via PRs of course) and publication would be via the subtree push method to gh-pages.

**Comment by billsacks** [comment]
Actually, I was just thinking... with the big caveat that I have just enough knowledge here to be dangerous: Isn't it possible now to have github.io generate web documentation based on your master branch? (I think your two choices are master and gh-pages.) If so, it feels like we should be able to do this without needing to have the extra level of indirection introduced by trying to use gh-pages for this. e.g., maybe having an index.html at the top level of cime, which redirects to the doc subdirectory, or something like that???

**Comment by gold2718** [comment]
In either case, the HTML generation has to happen offline and @rljacob makes the (valid IMHO) point that keeping all those generated HTML files in master is not a good idea (repo churn).
I think the doc **source** should end up in master with a publication procedure to gh-pages.

**Comment by rljacob** [comment]
One thing I noticed is that even if I do a fresh checkout and make no changes, "make html" will change files such that git notices.  Mostly "doctree" but some "html".   Maybe that's a version difference?  Anyway the only solution is to "git add ." everything under "build" so even more churn.

**Comment by agsalin** [comment]
It sounds to me that the solution to Rob's churn issue is to add all created files (*.html) to .gitignore so they don't end up in the repo at all.

**Comment by rljacob** [comment]
You have to push the html to github for it to display.  Github isn't smart enough to generate it (I don't think).

**Comment by agsalin** [comment]
Hmm. This doesn't sound scalable in that the repo will bloat over time. 

How about a separate repo that just has the html files? We can have a cron job that pulls master, regenerates them, and pushes them nightly to this other repo. The main CIME repo can just have a single html that links to the auto-generated repo. The cronjob can even get fancy, and add the current output of "create_newcase -- help" to the documentation.

**Comment by gold2718** [comment]
I can think of a variety of ways to curb repo bloat but the first step (as mentioned by @rljacob during the call) is to keep any generated files out of master.
Then we could use the gh-pages branch to manage bloat by clobbering old commits.

**Comment by billsacks** [comment]
In case you haven't seen them, some other relevant links:

http://nikhilism.com/post/2012/automatic-github-pages-generation-from/

https://daler.github.io/sphinxdoc-test/includeme.html

**Comment by gold2718** [comment]
@billsacks instead of just posting a bunch of links, could you summarize how they help advance the discussion? For instance, I don't see how the first link is relevant although it may be a useful addition to the [sphinx page on the wiki](https://github.com/ESMCI/cime/wiki/Getting-Started-with-Sphinx-and-reStructuredText).
As for the second link, is that a suggestion? A cautionary tale?

**Comment by billsacks** [comment]
@gold2718 - while I haven't had a chance to read them carefully, they both looked to me like solutions to this same problem - namely, people who wanted to do sphinx-based documentation in a subdirectory of their repo, without having the repo contain all of the generated html files.

(For context: I had been thinking, "Isn't it possible to to have github generate webpages based on rst files on the server?" In looking for the answer to that (which is apparently "no") I came across these pages that seemed relevant, so I thought I'd post them.)

**Comment by rljacob** [comment]
Another way to reduce bloat in the repo:  only re-generate the html every day or so.  Especially as we add a lot of content, it might be better to only commit changes to rst files. You should still build locally and check to make sure it works.  But don't push the html changes every time you make rst changes.

**Comment by rljacob** [comment]
NOTE:  to prevent repo bloat by removing commits, the .rst commits should be separate from the html commits.

**Comment by rljacob** [comment]
Closing this because I think we're all on the same page.
